### PR TITLE
Store the OCaml version used for Coq in vo files.

### PR DESCRIPTION
**Pull request by ppedrot**
See #12243.

### CODE REVIEW

**Review by ejgallego — CHANGES_REQUESTED** [review]
This solves little, and IMO is misguided; we shouldn't try to add incomplete patches and technical debt to workaround broken build systems.

In particular, linking errors can happen regardless of the OCaml version for example if using a development version or if any library or Coq module is updated.

**Comment by ppedrot** [comment]
@ejgallego I am fine with not merging this, it's just a POC.

**Comment by ejgallego** [comment]
> @ejgallego I am fine with not merging this, it's just a POC.

I think it solves little and on the contrary would add even more confusion, for example given two switches with the same compiler version, but I dunno, if people want this feel free to ignore my comment and merge.

**Comment by JasonGross** [comment]
Hm, I'm very much in support of something like this, but I don't think it makes much sense for it to happen on `Require`.  I think instead I want it to happen on `Declare ML Module`, and check the version of OCaml against the one in the compiled OCaml files.  Alternatively, or additionally, I think `coq_makefile` should generate `Makefile`s that issue warnings when trying to build OCaml files with a version of OCaml that does not match the one that compiled Coq.  (And there should be some similar warning or error that `dune` can emit, perhaps.)

**Comment by ejgallego** [comment]
Version of OCaml is close to irrelevant, an OCaml itself does the checks it needs. This is IMO thus a loss of time here in Coq; feel free to submit an improvement request to `Dynlink` upstream tho.

**Comment by ejgallego** [comment]
[p.s: My build system / setup of choice properly tracks dependencies so I have not seen this error since a long time]

**Comment by JasonGross** [comment]
> feel free to submit an improvement request to `Dynlink` upstream tho.

Sure, where's the issue tracker for it?

**Comment by ejgallego** [comment]
> Sure, where's the issue tracker for it?

https://github.com/ocaml/ocaml

**Comment by JasonGross** [comment]
Reported as https://github.com/ocaml/ocaml/issues/9532

**Comment by Zimmi48** [comment]
> In particular, linking errors can happen regardless of the OCaml version for example if using a development version or if any library or Coq module is updated.

What do you mean by this?

**Comment by ejgallego** [comment]
> What do you mean by this?

If we are talking about `cmxs` , as of today `Dynlink` will check the hash of all imported modules [I think, see my point, even myself I'm not sure of the exact rules, when they may change].

This can change under many conditions, not only the OCaml version. Dune for example can track this, opam is indeed being forced to rebuild everything.

My point is that the original error is invalid, in the sense the user is doing a non-supported thing and expecting this to be in the Coq manual? I dunno, I could for example update my `libc` and break C-level dynamic linking, still don't think Coq should document that. OCaml rules about having to rebuild everything are well known.

**Comment by JasonGross** [comment]
I have just seen the request of @cpitclaudel at https://github.com/coq/coq/pull/12243/files#r419137627; in light of this, I think storing the OCaml version in vo files makes sense, even if only for the ability to report it to the user.

**Comment by ppedrot** [comment]
Contrarily @ejgallego's comments, I don't believe that this PR is partial or introducing technical debt. Yet, if there is no consensus I'll close it soon because I don't like having lingering trivial PRs.

**Comment by ejgallego** [comment]
> Contrarily @ejgallego's comments, I don't believe that this PR is partial or introducing technical debt. Yet, if there is no consensus I'll close it soon because I don't like having lingering trivial PRs.

Indeed it is not so bad; as I wrote above I'm OK merging it. Will suggest a couple of improvements tho when I get a free minute.

File: vernac/library.ml [hunk 1]
```diff
@@ -401,6 +411,7 @@ let load_library_todo f =
   let tasks, _ = ObjFile.marshal_in_segment ch ~segment:"tasks" in
   let (s4 : seg_proofs), _ = ObjFile.marshal_in_segment ch ~segment:"opaques" in
   ObjFile.close_in ch;
+  check_ocaml_version s0.md_ocaml f;
```
Comment (ejgallego) [inline vernac/library.ml:414]: This duplication is annoying but there is little we can do about it.

**Review by ejgallego — COMMENTED** [review]
I have a bit of a gripe with the current PR due to the using the OCaml version as a fingerprint; IMO this is too weak and will give a very confusing or no error when a user gets to use two switches with the same version. IMHO I'd use the absolute path to the compiler + version + maybe arch (32/64) + magic numbers.

This could be encapsulated by adding a module `COCaml` so upper layers don't really care about the fingerprint details.

**Comment by ppedrot** [comment]
Absolute path is too strong, it makes the binary not relocatable. What magic numbers are you thinking about?

**Comment by ejgallego** [comment]
> Absolute path is too strong, it makes the binary not relocatable. What magic numbers are you thinking about?

What would be the use case where we relocate the OCaml compiler?

**Comment by ejgallego** [comment]
> What magic numbers are you thinking about?

Those on compiler-libs Config module, but indeed not sure if they are very useful.

**Comment by JasonGross** [comment]
> What would be the use case where we relocate the OCaml compiler?

When you want to install Coq on Windows to some drive other than C:\.  When two users have identical machines with identical setups but different usernames, and both have the same version of opam-installed Coq and OCaml, and they want to exchange vo files.  Using the absolute path is definitely way too strong to be an error, though I can see it being useful as a warning that can be turned off.

**Review by ejgallego — APPROVED** [review]
We are running out of time and it seems this check, while not perfect, would be useful in 8.12, so let's merge and refine later. I have two requests:
- move `check_ocaml_version` to a place in `lib/clib` so the message can be shared among coqchk / coq
- tweak a bit the message to better hint what the reality is, maybe add "recall that .vo files are only compatible when compiled with the exact same OCaml toolchain" or something like that?

**Comment by ppedrot** [comment]
@ejgallego done.

**Pull request merged**